File: lab/latency-dependency/probe.rb
require "socket"
require "json"

APP_HOST = ENV.fetch("APP_HOST", "app")
APP_PORT = Integer(ENV.fetch("APP_PORT", "8080"))
SAMPLES = Integer(ENV.fetch("SAMPLES", "15"))


def request(path)
  started = Process.clock_gettime(Process::CLOCK_MONOTONIC)
  socket = TCPSocket.new(APP_HOST, APP_PORT)
  socket.write("GET #{path} HTTP/1.1\r\nHost: #{APP_HOST}\r\nConnection: close\r\n\r\n")
  response = socket.read
  socket.close
  elapsed_ms = (Process.clock_gettime(Process::CLOCK_MONOTONIC) - started) * 1000.0
  body = response.split("\r\n\r\n", 2)[1].to_s
  [elapsed_ms, JSON.parse(body)]
end


def percentile50(values)
  sorted = values.sort
  sorted[sorted.length / 2]
end


def phase(delay_ms)
  request_latencies = []
  dependency_latencies = []
  SAMPLES.times do
    request_ms, body = request("/work?delay_ms=#{delay_ms}")
    request_latencies << request_ms
    dependency_latencies << body.fetch("dependency_latency_ms")
  end
  {
    samples: SAMPLES,
    configured_dependency_delay_ms: delay_ms,
    request_p50_ms: percentile50(request_latencies),
    dependency_p50_ms: percentile50(dependency_latencies)
  }
end

baseline = phase(5)
intervention = phase(120)
recovery = phase(5)

request_delta = intervention[:request_p50_ms] - baseline[:request_p50_ms]
dependency_delta = intervention[:dependency_p50_ms] - baseline[:dependency_p50_ms]
recovery_request_delta = (recovery[:request_p50_ms] - baseline[:request_p50_ms]).abs
propagation_error = (request_delta - dependency_delta).abs

thresholds = {
  min_request_delta_ms: 80.0,
  min_dependency_delta_ms: 80.0,
  max_recovery_request_delta_ms: 30.0,
  max_propagation_error_ms: 30.0
}

assertions = {
  dependency_latency_increased: dependency_delta >= thresholds[:min_dependency_delta_ms],
  upstream_latency_increased: request_delta >= thresholds[:min_request_delta_ms],
  recovery_returned_toward_baseline: recovery_request_delta <= thresholds[:max_recovery_request_delta_ms],
  downstream_delay_explains_upstream_delta: propagation_error <= thresholds[:max_propagation_error_ms]
}

result = assertions.values.all? ? "supports" : "contradicts"

evidence = {
  schema_version: "0.1",
  kind: "empirical_evidence",
  experiment: "experiment.latency.external_dependency.ruby_http",
  claims: ["claim.latency.external_dependency.delay_propagates_upstream"],
  environment: {runtime: "ruby", runtime_version: RUBY_VERSION, platform: RUBY_PLATFORM, isolation: "docker_compose"},
  intervention: {action: "increase_downstream_dependency_delay", baseline_delay_ms: 5, intervention_delay_ms: 120, recovery_delay_ms: 5},
  observations: {
    baseline: baseline,
    intervention: intervention,
    recovery: recovery,
    derived: {request_latency_delta_ms: request_delta, dependency_latency_delta_ms: dependency_delta, recovery_request_delta_ms: recovery_request_delta, propagation_error_ms: propagation_error},
    thresholds: thresholds
  },
  assertions: assertions,
  result: result,
  interpretation: "A controlled delay added only to a synchronous downstream HTTP dependency should appear both in measured dependency latency and in upstream end-to-end request latency, then disappear when the delay is removed.",
  limitations: [
    "The application and dependency are synthetic Ruby services on one Docker host.",
    "This validates critical-path latency propagation, not the cause of latency inside a real third-party provider.",
    "Parallel calls, retries, timeouts, caching, asynchronous work, and queueing can change propagation behavior."
  ]
}

puts JSON.generate(evidence)

File: lab/latency-database/probe.py
import http.client
import json
import os
import platform
import statistics
import time

APP_HOST = os.environ.get("APP_HOST", "app")
APP_PORT = int(os.environ.get("APP_PORT", "8080"))
SAMPLES = int(os.environ.get("SAMPLES", "12"))
EXPERIMENT_ID = "experiment.latency.database.query_delay_python_postgres"
CLAIM_ID = "claim.latency.database.query_delay_propagates_upstream"


def one_request(delay_ms: int):
    started = time.monotonic()
    connection = http.client.HTTPConnection(APP_HOST, APP_PORT, timeout=10)
    connection.request("GET", f"/work?delay_ms={delay_ms}")
    response = connection.getresponse()
    body = response.read()
    connection.close()
    elapsed_ms = (time.monotonic() - started) * 1000.0
    if response.status != 200:
        raise RuntimeError(f"app returned {response.status}: {body.decode(errors='replace')}")
    payload = json.loads(body)
    return elapsed_ms, float(payload["database_query_latency_ms"])


def phase(delay_ms: int):
    request_latencies = []
    database_latencies = []
    for _ in range(SAMPLES):
        request_ms, db_ms = one_request(delay_ms)
        request_latencies.append(request_ms)
        database_latencies.append(db_ms)
    return {
        "configured_database_delay_ms": delay_ms,
        "database_query_p50_ms": statistics.median(database_latencies),
        "request_p50_ms": statistics.median(request_latencies),
        "samples": SAMPLES,
    }


baseline = phase(0)
intervention = phase(120)
recovery = phase(0)

db_delta = intervention["database_query_p50_ms"] - baseline["database_query_p50_ms"]
request_delta = intervention["request_p50_ms"] - baseline["request_p50_ms"]
propagation_error = abs(request_delta - db_delta)
recovery_request_delta = abs(recovery["request_p50_ms"] - baseline["request_p50_ms"])

thresholds = {
    "min_database_delta_ms": 80.0,
    "min_request_delta_ms": 80.0,
    "max_propagation_error_ms": 35.0,
    "max_recovery_request_delta_ms": 35.0,
}
assertions = {
    "database_latency_increased": db_delta >= thresholds["min_database_delta_ms"],
    "upstream_latency_increased": request_delta >= thresholds["min_request_delta_ms"],
    "database_delay_explains_upstream_delta": propagation_error <= thresholds["max_propagation_error_ms"],
    "recovery_returned_toward_baseline": recovery_request_delta <= thresholds["max_recovery_request_delta_ms"],
}
result = "supports" if all(assertions.values()) else "inconclusive"

evidence = {
    "schema_version": "0.1",
    "kind": "empirical_evidence",
    "experiment": EXPERIMENT_ID,
    "claims": [CLAIM_ID],
    "environment": {
        "runtime": "python",
        "runtime_version": platform.python_version(),
        "platform": platform.platform(),
        "isolation": "docker_compose",
    },
    "intervention": {
        "action": "increase_postgresql_query_delay",
        "mechanism": "pg_sleep",
        "baseline_delay_ms": 0,
        "intervention_delay_ms": 120,
        "recovery_delay_ms": 0,
    },
    "observations": {
        "baseline": baseline,
        "intervention": intervention,
        "recovery": recovery,
        "derived": {
            "database_latency_delta_ms": db_delta,
            "request_latency_delta_ms": request_delta,
            "propagation_error_ms": propagation_error,
            "recovery_request_delta_ms": recovery_request_delta,
        },
        "thresholds": thresholds,
    },
    "assertions": assertions,
    "result": result,
    "interpretation": "A controlled delay added only to a synchronous PostgreSQL query should appear in both measured database-call latency and upstream end-to-end request latency, then disappear when the query delay is removed.",
    "limitations": [
        "The database and application are synthetic services on one Docker host.",
        "pg_sleep validates database-side query delay propagation but not execution-plan, lock, storage, or network causes.",
        "The app launches psql per request, adding connection and process startup overhead to every phase.",
    ],
}
print(json.dumps(evidence))
